coderd/notifications/manager.go

```diff
@@ -0,0 +1,450 @@
+		// see BulkMarkNotificationMessagesSent/BulkMarkNotificationMessagesFailed. If we had the ability to batch updates,
+		// like is offered in https://docs.sqlc.dev/en/stable/reference/query-annotations.html#batchmany, we'd have a cleaner
+		// approach to this - but for now this will work fine.
+		success = make(chan dispatchResult, m.cfg.StoreSyncBufferSize)
+		failure = make(chan dispatchResult, m.cfg.StoreSyncBufferSize)
+	)
+
+	// Create a specific number of notifiers to run concurrently.
+	var eg errgroup.Group
+	for i := 0; i < notifiers; i++ {
+		eg.Go(func() error {
+			m.notifierMu.Lock()
+			n := newNotifier(ctx, m.cfg, i+1, m.log, m.store, m.handlers)
+			m.notifiers = append(m.notifiers, n)
+			m.notifierMu.Unlock()
+			return n.run(ctx, success, failure)
+		})
+	}
+
+	eg.Go(func() error {
+		// Every interval, collect the messages in the channels and bulk update them in the database.
+		tick := time.NewTicker(m.cfg.StoreSyncInterval.Value())
+		defer tick.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				// Nothing we can do in this scenario except bail out; after the message lease expires, the messages will
+				// be requeued and users will receive duplicates.
+				// This is an explicit trade-off between keeping the database load light (by bulk-updating records) and
+				// exactly-once delivery.
+				//
+				// The current assumption is that duplicate delivery of these messages is, at worst, slightly annoying.
+				// If these notifications are triggering external actions (e.g. via webhooks) this could be more
+				// consequential, and we may need a more sophisticated mechanism.
+				//
+				// TODO: mention the above tradeoff in documentation.
+				m.log.Warn(ctx, "exiting ungracefully", slog.Error(ctx.Err()))
+
+				if len(success)+len(failure) > 0 {
+					m.log.Warn(ctx, "content canceled with pending updates in buffer, these messages will be sent again after lease expires",
+						slog.F("success_count", len(success)), slog.F("failure_count", len(failure)))
+				}
+				return ctx.Err()
+			case <-m.stop:
+				if len(success)+len(failure) > 0 {
+					m.log.Warn(ctx, "flushing buffered updates before stop",
+						slog.F("success_count", len(success)), slog.F("failure_count", len(failure)))
+					m.bulkUpdate(ctx, success, failure)
+					m.log.Warn(ctx, "flushing updates done")
+				}
+				return nil
+			case <-tick.C:
+				m.bulkUpdate(ctx, success, failure)
+			}
+		}
+	})
+
+	err := eg.Wait()
+	if err != nil {
+		m.log.Error(ctx, "manager loop exited with error", slog.Error(err))
+	}
+	return err
+}
+
+// Enqueue queues a notification message for later delivery.
+// Messages will be dequeued by a notifier later and dispatched.
+func (m *Manager) Enqueue(ctx context.Context, userID, templateID uuid.UUID, labels types.Labels, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
+	payload, err := m.buildPayload(ctx, userID, templateID, labels)
```

Comment:
> Why do we need to compute the payload before the message goes into the database?

I like the separation of concerns here. The enqueuing process prepares the payload, and the dispatch process renders the template based on the payload. The side-effect of this is also that we build the payload once, instead of on each dispatch attempt.

> If you didn't need to compute the payload here, then you wouldn't need a Manager to enqueue notifications, we could have a very lightweight "enqueuer" that just needs a database.Store. I think that would alleviate some of the plumbing that motivated you to have a singleton.
> 
> But, even if we do need to compute the payload prior to enqueuing, I think Manager should be split into enqueueing and dispatching. No consumer of this package needs to do both things at the same time, and they don't share any state.

This is a great call. I implemented this in 884fadf9469962645ae46e60b71f02e3905b2a6f; was an easy refactor and feels much cleaner.